File: Java/src/main/java/com/contidio/sdk/exceptions/BackendException.java
package com.contidio.sdk.exceptions;

import org.apache.commons.lang3.StringUtils;

import com.contidio.sdk.proto.MetadataProtocol.BackendErrorCode;
import com.contidio.sdk.proto.MetadataProtocol.Error;

@SuppressWarnings("serial")
public class BackendException extends Exception {
	private BackendErrorCode errorCode;	
	private String transientClass;
	private int httpErrorCode;
	
	public BackendException(BackendErrorCode errorCode, String message, String transientClass, int httpErrorCode) {
		super(message);
		
		this.errorCode = errorCode;		
		this.transientClass = transientClass;
		this.httpErrorCode = httpErrorCode;
	}
	
	public BackendErrorCode getErrorCode() {
		return errorCode;
	}
	
	public String getTransientClass() {
		return transientClass;
	}
		
	public int getHttpErrorCode() {
		return httpErrorCode;
	}

	public static Exception getTransientExceptionFor(Error error) {
		String exceptionClass = error.getExceptionType();
        if (StringUtils.equals(exceptionClass, "AlreadyExistsBackendException"))
            return new AlreadyExistsBackendException(error.getBackendErrorCode(), error.getErrorMessage());
        else if (StringUtils.equals(exceptionClass, "NotFoundBackendException"))
            return new NotFoundBackendException(error.getBackendErrorCode(), error.getErrorMessage());
        else if (StringUtils.equals(exceptionClass, "InsufficientPrivilegesBackendException"))
            return new InsufficientPrivilegesBackendException();
        else if (StringUtils.equals(exceptionClass, "InvalidCredentialsBackendException"))
            return new InvalidCredentialsBackendException();
        else if (StringUtils.equals(exceptionClass, "InternalErrorBackendException"))
            return new InternalErrorBackendException(error.getErrorMessage());
        else if (StringUtils.equals(exceptionClass, "InvalidOperationBackendException"))
            return new InvalidOperationBackendException(error.getBackendErrorCode(), error.getErrorMessage());
        else if (StringUtils.equals(exceptionClass, "OptimisticLockingBackendException"))
            return new OptimisticLockingBackendException(error.getErrorMessage());
        else if (StringUtils.equals(exceptionClass, "ServiceUnavailableBackendException"))
            return new ServiceUnavailableBackendException(error.getErrorMessage());
        else if (StringUtils.equals(exceptionClass, "PostponeExecutionBackendException"))
            return new PostponeExecutionBackendException();
        else if (StringUtils.equals(exceptionClass, "StopExecutionBackendException"))
            return new StopExecutionBackendException();
        else
            return new ServiceUnavailableBackendException("Invalid error data received from service: " +
                error.getErrorMessage());       
	}
}

File: .NET Core/Exceptions/BackendException.cs
using com.contidio.sdk.proto;
using System;
using System.Net;

namespace Contidio.Sdk.Exceptions
{
    public class BackendException : Exception
    {
        public BackendErrorCode ErrorCode { get; private set; }
        public String TransientClass { get; private set; }
        public HttpStatusCode HttpErrorCode { get; private set; }

        public BackendException(BackendErrorCode errorCode, String message, String transientClass, HttpStatusCode httpErrorCode) 
            : base(message)
        {
            ErrorCode = errorCode;
            TransientClass = transientClass;
            HttpErrorCode = httpErrorCode;
        }

        public static Exception GetTransientExceptionFor(Error error)
        {
            string exceptionClass = error.ExceptionType;
            if (string.Equals(exceptionClass, "AlreadyExistsBackendException"))
                return new AlreadyExistsBackendException(error.BackendErrorCode, error.ErrorMessage);
            else if (string.Equals(exceptionClass, "NotFoundBackendException"))
                return new NotFoundBackendException(error.BackendErrorCode, error.ErrorMessage);
            else if (string.Equals(exceptionClass, "InsufficientPrivilegesBackendException"))
                return new InsufficientPrivilegesBackendException();
            else if (string.Equals(exceptionClass, "InvalidCredentialsBackendException"))
                return new InvalidCredentialsBackendException();
            else if (string.Equals(exceptionClass, "InternalErrorBackendException"))
                return new InternalErrorBackendException(error.ErrorMessage);
            else if (string.Equals(exceptionClass, "InvalidOperationBackendException"))
                return new InvalidOperationBackendException(error.BackendErrorCode, error.ErrorMessage);
            else if (string.Equals(exceptionClass, "OptimisticLockingBackendException"))
                return new OptimisticLockingBackendException(error.ErrorMessage);
            else if (string.Equals(exceptionClass, "ServiceUnavailableBackendException"))
                return new ServiceUnavailableBackendException(error.ErrorMessage);
            else if (string.Equals(exceptionClass, "PostponeExecutionBackendException"))
                return new PostponeExecutionBackendException();
            else if (string.Equals(exceptionClass, "StopExecutionBackendException"))
                return new StopExecutionBackendException();
            else
                return new ServiceUnavailableBackendException("Invalid error data received from service: " +
                    error.ErrorMessage);            
        }
    }
}
